ui/app/components/inference/ToolParametersSection.tsx

```diff
@@ -0,0 +1,278 @@
+                    {parallelToolCalls ? "enabled" : "disabled"}
+                  </Badge>
+                </div>
+              )}
+            </div>
+          )}
+
+          {/* Allowed Tools */}
+          {hasAllowedTools && (
+            <div className="flex flex-col gap-2">
+              <span className="text-fg-muted text-xs font-medium uppercase">
+                Allowed Tools ({allowedTools.length})
+              </span>
+              <div className="flex flex-wrap gap-2">
+                {allowedTools.map((toolName) => (
+                  <Badge key={toolName} variant="outline">
+                    {toolName}
+                  </Badge>
+                ))}
+              </div>
+            </div>
+          )}
+
+          {/* Additional Tools */}
+          {hasAdditionalTools && (
+            <div className="flex flex-col gap-2">
+              <span className="text-fg-muted text-xs font-medium uppercase">
+                Additional Tools ({additionalTools.length})
+              </span>
+              <AdditionalToolsList tools={additionalTools} />
+            </div>
+          )}
+
+          {/* Provider Tools */}
+          {hasProviderTools && (
+            <div className="flex flex-col gap-2">
+              <span className="text-fg-muted text-xs font-medium uppercase">
+                Provider Tools ({providerTools.length})
+              </span>
+              <ProviderToolsList tools={providerTools} />
+            </div>
+          )}
+        </div>
+      </SnippetContent>
+    </SnippetLayout>
+  );
+}
+
+function ToolChoiceBadge({ toolChoice }: { toolChoice: ToolChoice }) {
+  if (typeof toolChoice === "string") {
+    return <Badge variant="outline">{toolChoice}</Badge>;
+  }
+
+  // toolChoice is { specific: string }
+  return (
+    <Badge variant="outline">
+      specific: <span className="font-mono">{toolChoice.specific}</span>
+    </Badge>
+  );
+}
+
+function AdditionalToolsList({ tools }: { tools: Tool[] }) {
+  if (tools.length === 1) {
+    return <ToolCard tool={tools[0]} />;
+  }
+
+  const tabs = tools.map((tool, index) => ({
+    id: String(index),
+    label: getToolName(tool),
+    content: <ToolCard tool={tool} />,
+  }));
+
+  return <SnippetTabs tabs={tabs} />;
+}
+
+function getToolName(tool: Tool): string {
+  if (tool.type === "function") {
+    return tool.name;
+  } else if (tool.type === "openai_custom") {
+    return tool.name;
+  }
+  return "Unknown Tool";
+}
+
+function ToolCard({ tool }: { tool: Tool }) {
+  if (tool.type === "function") {
+    return <FunctionToolCard tool={tool} />;
+  } else if (tool.type === "openai_custom") {
+    return <OpenAICustomToolCard tool={tool} />;
+  }
+  return null;
+}
+
+type FunctionToolWithType = { type: "function" } & FunctionTool;
+type OpenAICustomToolWithType = { type: "openai_custom" } & OpenAICustomTool;
+
+function FunctionToolCard({ tool }: { tool: FunctionToolWithType }) {
```

Comment:
nit: I think you can assign an object like `{ type: "function" } & FunctionTool` to a type expecting `FunctionTool` so these `ToolWithType` types shouldn't be necessary?